minio/credentials/providers.py

```diff
@@ -508,18 +508,13 @@ def retrieve(self) -> Credentials:
             res = _urlopen(
                 self._http_client,
                 "GET",
-                urlunsplit(
-                    url_replace(
-                        urlsplit(url),
-                        path="/latest/meta-data/iam/security-credentials/",
-                    ),
-                ),
+                url+"/latest/meta-data/iam/security-credentials/",
```

Comment:
This breaks if `custom_endpoint` is passed. I think you would need to leave it as is.

Reply:
Thanks for the feedback!

I noticed the same logic is already applied on 
https://github.com/minio/minio-py/blob/759a5f3f432e884823d040e1746600c537ad6c41/minio/credentials/providers.py#L501

which follows the same pattern used in the Go implementation 
https://github.com/minio/minio-go/blob/ab1c2fe050a4dd502c863178696a54c44610ca82/pkg/credentials/iam_aws.go#L334

However, the Go implementation has an issue: it doesn’t account for cases where a custom endpoint contains a path. This leads to the assumption that the custom_endpoint doesn’t have a path, which is why I addressed it this way on line 511.

To resolve this, we have two options when a custom_endpoint is passed:

1. **Assume it does not contain a path** and use simple string concatenation (`+`) in all instances where we manipulate the URL.
2. **Assume it can contain a path** and ensure proper handling by using urllib.parse.SplitResult in all usages to safely override the path.

Given that the Go implementation itself is inconsistent—sometimes appending directly to the URL and other times parsing the URL—it would be helpful to align on one approach to avoid further confusion and potential issues.

Which approach do you think would be best to adopt moving forward?

Reply:
Thanks for the feedback!

I noticed the same logic is already applied on 
https://github.com/minio/minio-py/blob/759a5f3f432e884823d040e1746600c537ad6c41/minio/credentials/providers.py#L501

which follows the same pattern used in the Go implementation 
https://github.com/minio/minio-go/blob/ab1c2fe050a4dd502c863178696a54c44610ca82/pkg/credentials/iam_aws.go#L334

However, the Go implementation has an issue: it doesn’t account for cases where a custom endpoint contains a path. This leads to the assumption that the custom_endpoint doesn’t have a path, which is why I addressed it this way on line 511.

To resolve this, we have two options when a custom_endpoint is passed:

1. **Assume it does not contain a path** and use simple string concatenation (`+`) in all instances where we manipulate the URL.
2. **Assume it can contain a path** and ensure proper handling by using urllib.parse.SplitResult in all usages to safely override the path.

Given that the Go implementation itself is inconsistent—sometimes appending directly to the URL and other times parsing the URL—it would be helpful to align on one approach to avoid further confusion and potential issues.

Which approach do you think would be best to adopt moving forward?